spring-integration-core/src/main/java/org/springframework/integration/util/FunctionIterator.java

```diff
@@ -64,6 +106,27 @@ public void close() throws IOException {
 		if (this.iterator instanceof Closeable) {
 			((Closeable) this.iterator).close();
 		}
+		else if (this.iterator instanceof AutoCloseable) {
+			try {
+				((AutoCloseable) this.iterator).close();
+			}
+			catch (Exception e) {
+				// NOSONAR
+			}
+		}
+		if (!this.root.equals(this.iterator)) {
+			if (this.root instanceof Closeable) {
```

Comment:
See comment in JIRA: https://jira.spring.io/browse/INT-4506?focusedCommentId=161106&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-161106